### ISSUE TITLE
CompositeQualifiers reuses stream

### ISSUE THREAD

**Issue by musketyr** [micronaut-projects/micronaut-core#679]
`CompositeQualifier` made by `Qualifiers.byQualifiers` does not work at the moment as it reduces the same stream causing the following exception:

```
	at com.agorapulse.micronaut.grails.QualifiersSpec.qualifiers does not work(QualifiersSpec.groovy:20)
Caused by: java.lang.IllegalStateException: stream has already been operated upon or closed
	at java.util.stream.AbstractPipeline.<init>(AbstractPipeline.java:203)
	at java.util.stream.ReferencePipeline.<init>(ReferencePipeline.java:94)
	at java.util.stream.ReferencePipeline$StatelessOp.<init>(ReferencePipeline.java:618)
	at java.util.stream.ReferencePipeline$2.<init>(ReferencePipeline.java:163)
	at java.util.stream.ReferencePipeline.filter(ReferencePipeline.java:162)
	at io.micronaut.inject.qualifiers.NameQualifier.reduce(NameQualifier.java:56)
	at io.micronaut.inject.qualifiers.CompositeQualifier.reduce(CompositeQualifier.java:50)
	at io.micronaut.context.DefaultBeanContext.getBeanDefinitions(DefaultBeanContext.java:805)
```

### Steps to Reproduce

```
import io.micronaut.context.DefaultBeanContext
import io.micronaut.context.Qualifier
import io.micronaut.inject.qualifiers.Qualifiers
import spock.lang.Specification

import javax.inject.Named
import javax.inject.Singleton

class QualifiersSpec extends Specification {

    void 'qualifiers does not work'() {
        given:
            DefaultBeanContext context = new DefaultBeanContext()
            context.start()
        when:
            Qualifier qualifier = Qualifiers.byQualifiers(Qualifiers.byType(Runnable), Qualifiers.byName('thread'))
        then:
            context.getBeanDefinitions(qualifier)

    }

}

@Singleton
@Named('thread')
class TestMe implements Runnable {
    @Override
    void run() {

    }
}
```

### Expected Behaviour

The test passes.

### Actual Behaviour

Exception is thrown instead

### Environment Information
- **Micronaut Version:** 1.0.0.RC1
- **JDK Version:** 1.8

**Comment by jameskleeh** [comment]
@musketyr I've added tests to `inject-java` and `inject-groovy` that follows your example and the tests pass. https://github.com/micronaut-projects/micronaut-core/commit/c96ee33ef21df3ecfd785ea066e1328aa6618e82

If you can tell me what needs to be different about the test to cause it to fail, I'd be happy to take another look.

**Comment by graemerocher** [comment]
I believe I fixed this in master with 2c62b4afbd09198cb51605074761e61a4e4216f0

Thanks for the test

**Comment by musketyr** [comment]
Yes @graemerocher, I can see a fix in there. Thank you @jameskleeh for the tests and for thanks both for fast response!
